### PR TITLE
Update auth example with type narrowing

### PR DESCRIPTION
**Pull request by larsenwork**
Should limit issues/requests like 
* #8503
* #8522
* #8515

### CODE REVIEW

File: examples/auth/src/getPathname.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { Location } from "react-router"
+
+const isObjectWithKey = <T extends string>(
+  given: unknown,
+  key: T
+): given is Partial<Record<T, unknown>> =>
+  typeof given === 'object' && given !== null && key in given
+
+export const getPathname = (location: Location): string | undefined => {
```
Comment (larsenwork) [inline examples/auth/src/getPathname.ts:9]: Some variant of this function could also exist in the package itself for better community reuse but I wanted to POC it in the example first.

File: examples/auth/src/getPathname.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Location } from "react-router"
+
+const isObjectWithKey = <T extends string>(
```
Comment (larsenwork) [inline examples/auth/src/getPathname.ts:3]: Using something like Lodash `get` could replace this but I didn't want to rely on a dependency.

**Comment by timdorr** [comment]
Thanks, but I don't think this will necessarily address that concern. This is just an example and in absence of folks copying out the example `getPathname` code they will still end up with an `unknown` type in most cases. 

The more likely issue is when someone destructures `location.state` directly into a variable and gets a type error. E.g., `const { foo: { bar } } = location.state`. That's still going to be the case because of the underlying types.

**Comment by larsenwork** [comment]
Unsure I follow. In the current form the auth example doesn't even lint? This PR would show users how to consume useLocation in a type safe manor (that also lints).

image

**Comment by cainlevy** [comment]
Without a concrete example and recommendation of how to do this the intended way, I think most people will cast the value. It's pretty easy when the return type is `unknown` but it leads to a worse outcome than the previous generic:

```
const location = useLocation() as MyLocationType;
```

**Pull request closed without merging**